### PR TITLE
Feedback

### PR DESCRIPTION
**Pull request by BoriskinaAnna**

### CODE REVIEW

File: ITechArt.StudentsLab.BusinessLayer/Contracts/IFeedbackService.cs [hunk 1]
```diff
@@ -6,14 +6,20 @@ namespace ITechArt.StudentsLab.BusinessLayer.Contracts
 {
     public interface IFeedbackService
     {
-        Task<IEnumerable<FeedbackDateModel>> GetFeedbackDates(int labId);
+        Task<IEnumerable<FeedbackDateResponseModel>> GetFeedbackDates(int labId);
```
Comment (andrei-shenets-simpligov) [inline ITechArt.StudentsLab.BusinessLayer/Contracts/IFeedbackService.cs]: Do you really need separate FeedbackDateResponseModel and FeedbackDateRequestModel instead of just FeedbackDateModel?

File: ITechArt.StudentsLab.BusinessLayer/Contracts/IFeedbackService.cs [hunk 2]
```diff
@@ -6,14 +6,20 @@ namespace ITechArt.StudentsLab.BusinessLayer.Contracts
...
 
-        Task<FeedbackDateModel> UpsertFeedbackDate(FeedbackDateModel feedbackDate);
+        Task<UpsertResponseModel> UpsertFeedbackDate(FeedbackDateRequestModel feedbackDate);
```
Comment (andrei-shenets-simpligov) [inline ITechArt.StudentsLab.BusinessLayer/Contracts/IFeedbackService.cs]: I recommend to use response and request only on presentation level. In this case I would use UpsertResultModel or Just UpsertResult. Let's discuss

File: ITechArt.StudentsLab.BusinessLayer/Contracts/IFeedbackService.cs [hunk 3]
```diff
@@ -6,14 +6,20 @@ namespace ITechArt.StudentsLab.BusinessLayer.Contracts
...
         Task<IEnumerable<FeedbackAnswerResponseModel>> GetFeedbackAnswers(FeedbackAnswerGetRequestModel feedbackRequest);
 
         Task<IEnumerable<FeedbackQuestionModel>> GetFeedbackQuestions(int labId);
 
-        Task UpsertFeedbackAnswer(FeedbackAnswerPostRequestModel feedbackAnswer);
+        Task<UpsertResponseModel> UpsertFeedbackAnswer(FeedbackAnswerPostRequestModel feedbackAnswer);
+
+        Task<int> UpsertFeedbackQuestion(FeedbackQuestionRequestModel feedbackQuestion);
+
+        void DeleteFeedbackQuestion(int questionId, int labId);
```
Comment (andrei-shenets-simpligov) [inline ITechArt.StudentsLab.BusinessLayer/Contracts/IFeedbackService.cs]: Why this method is not async?

File: ITechArt.StudentsLab.BusinessLayer/Contracts/IFeedbackService.cs [hunk 4]
```diff
@@ -6,14 +6,20 @@ namespace ITechArt.StudentsLab.BusinessLayer.Contracts
 {
     public interface IFeedbackService
     {
-        Task<IEnumerable<FeedbackDateModel>> GetFeedbackDates(int labId);
+        Task<IEnumerable<FeedbackDateResponseModel>> GetFeedbackDates(int labId);
 
-        Task<FeedbackDateModel> UpsertFeedbackDate(FeedbackDateModel feedbackDate);
+        Task<UpsertResponseModel> UpsertFeedbackDate(FeedbackDateRequestModel feedbackDate);
 
         Task<IEnumerable<FeedbackAnswerResponseModel>> GetFeedbackAnswers(FeedbackAnswerGetRequestModel feedbackRequest);
```
Comment (andrei-shenets-simpligov) [inline ITechArt.StudentsLab.BusinessLayer/Contracts/IFeedbackService.cs]: All async methods should have Async postfix

File: ITechArt.StudentsLab.DataAccessLayer/Repositories/FeedbackRepository.cs [hunk 5]
```diff
@@ -5,104 +5,153 @@
...
+                await connection.ExecuteAsync(
+                    "DeleteFeedbackQuestion",
+                    new { Id = questionId },
+                    commandType: CommandType.StoredProcedure
+                );
             }
         }
 
-        public async Task<int> UpsertFeedbackDates(FeedbackDate feedbackDate)
+        public async void DeleteFeedbackDate(int dateId)
```
Comment (andrei-shenets-simpligov) [inline ITechArt.StudentsLab.DataAccessLayer/Repositories/FeedbackRepository.cs]: Incorrect implementation

File: ITechArt.StudentsLab.Web/Controllers/FeedbacksController.cs [hunk 6]
```diff
@@ -22,59 +22,77 @@ public FeedbacksController(IFeedbackService feedbackService)
...
+
+        [HttpDelete("labs/feedbacks/dates/{dateId:int}")]
+        public void DeleteFeedbackDate(int dateId)
+        {
+            _feedbackService.DeleteFeedbackDate(dateId);
         }
 
-        [HttpGet("labs/feedbacks/dates/{feedbackDateId:int}/mentors/{mentorId:int}/students/{studentId:int}")]
-        public async Task<IActionResult> GetFeedbackAnswers(int studentId, int mentorId, int feedbackDateId)
+        [HttpGet("labs/feedbacks/dates/{feedbackDateId:int}/mentors/students/{studentId:int}")]
```
Comment (andrei-shenets-simpligov) [inline ITechArt.StudentsLab.Web/Controllers/FeedbacksController.cs]: I think we can simplify this route. Let's discuss.

File: front/.eslintrc.json [hunk 7]
```diff
@@ -61,7 +61,7 @@
       "warn"
     ],
     "no-console": [
-      "error",
+      "warn",
```
Comment (andrei-shenets-simpligov) [inline front/.eslintrc.json]: why did you change this?

File: front/.eslintrc.json [hunk 8]
```diff
@@ -137,7 +136,7 @@
       "always"
     ],
     "array-element-newline": [
-      "error",
+      "warn",
```
Comment (andrei-shenets-simpligov) [inline front/.eslintrc.json]: why did you change this?

File: front/src/components/AdminFeedback/index.js [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+
+class AdminFeedback extends Component {
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      'questions': [],
+      'dates': [],
+      'isQuestionsLoading': true,
```
Comment (andrei-shenets-simpligov) [inline front/src/components/AdminFeedback/index.js]: why do you need quotes for property names?

File: front/src/components/AdminFeedback/index.js [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+            this.upserQuestionLocal(result.data);
+          }
+        })
+        .catch(error => console.error(error));
+    }
+  };
+
+  upsertDateLocal = (dateId) => {
+    const dates = this.state.dates;
+    const changeIndex = dates.findIndex(date => date.id === dateId);
```
Comment (andrei-shenets-simpligov) [inline front/src/components/AdminFeedback/index.js]: I think find() method is better than findIndex().

File: front/src/components/AdminFeedback/index.js [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+      'dates': dates,
+      'isAddingDate': false,
+      'currentDateId': undefined
+    });
+  };
+
+  upsertDate = (dateId) => {
+    const addedDate = new Date(this.state.calendarValue);
+    const existingValueIndex = this.state.dates.findIndex(value => {
+      const savedDate = new Date(value.date);
```
Comment (andrei-shenets-simpligov) [inline front/src/components/AdminFeedback/index.js]: what form value has? what format has value.date? why do you need to create new Date from it?

File: front/src/components/AdminFeedback/index.js [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+    this.setState({
+      'dates': dates,
+      'isAddingDate': false,
+      'currentDateId': undefined
+    });
+  };
+
+  upsertDate = (dateId) => {
+    const addedDate = new Date(this.state.calendarValue);
+    const existingValueIndex = this.state.dates.findIndex(value => {
```
Comment (andrei-shenets-simpligov) [inline front/src/components/AdminFeedback/index.js]: fin()

File: front/src/languages/lang.json [hunk 13]
```diff
@@ -35,7 +35,30 @@
       "phSearchUser":"Search user",
       "btnAddUser":"Add user",
       "lbUsersHeader":"Users",
-      "lbMonths":["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
+      "lbMonths":["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"],
+      "addQuestion": "Add question",
+      "jan": "Jan",
```
Comment (andrei-shenets-simpligov) [inline front/src/languages/lang.json]: Not the best solution. Let's discuss

File: front/src/services/fetchService/fetchResultTypeEnum.js [hunk 14]
```diff
@@ -5,5 +5,9 @@ export const FetchResultTypeEnum = {
 
   'INTERNAL_SERVER_ERROR': 3,
 
-  'SUCCESS': 4
+  'SUCCESS': 4,
+
+  'ALREADY_EXISTS_ERROR': 5,
+
+  'BAD_REQUEST': 6
```
Comment (andrei-shenets-simpligov) [inline front/src/services/fetchService/fetchResultTypeEnum.js]: Why do we have this? We have http status 400 for this.

File: front/src/services/fetchService/fetchResultTypeEnum.js [hunk 15]
```diff
@@ -5,5 +5,9 @@ export const FetchResultTypeEnum = {
 
   'INTERNAL_SERVER_ERROR': 3,
 
-  'SUCCESS': 4
+  'SUCCESS': 4,
+
+  'ALREADY_EXISTS_ERROR': 5,
```
Comment (andrei-shenets-simpligov) [inline front/src/services/fetchService/fetchResultTypeEnum.js]: Why do we have this? We have http status 409 for this.

File: front/src/services/fetchService/fetchResultTypeEnum.js [hunk 16]
```diff
@@ -5,5 +5,9 @@ export const FetchResultTypeEnum = {
 
   'INTERNAL_SERVER_ERROR': 3,
 
-  'SUCCESS': 4
+  'SUCCESS': 4,
```
Comment (andrei-shenets-simpligov) [inline front/src/services/fetchService/fetchResultTypeEnum.js]: Why do we have this? We have http statuses 2XX for this.

File: front/src/services/fetchService/fetchResultTypeEnum.js [hunk 17]
```diff
@@ -5,5 +5,9 @@ export const FetchResultTypeEnum = {
 
   'INTERNAL_SERVER_ERROR': 3,
```
Comment (andrei-shenets-simpligov) [inline front/src/services/fetchService/fetchResultTypeEnum.js]: Why do we have this? We have http status 500 for this.

File: front/src/components/AdminFeedback/index.js [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+                && addedDate.getMonth() === savedDate.getMonth()
+                && addedDate.getDate() === savedDate.getDate();
+    });
+    if (existingValue) {
+      this.setState({ existingDateError: true });
+    } else {
+      feedbackService.upsertDate(this.state.calendarValue, this.props.labId, dateId)
+        .then(result => {
+          switch (result.type) {
+          case HttpStatus.CONFLICT: this.setState({ existingDateError: true });
```
Comment (AndreiShenets) [inline front/src/components/AdminFeedback/index.js]: Do not put statements after case. 
Indend case statements once

File: front/src/components/Schedule/index.js [hunk 19]
```diff
@@ -18,7 +23,8 @@ class Schedule extends Component {
       'submenu': 'lectures',
```
Comment (AndreiShenets) [inline front/src/components/Schedule/index.js]: Quotes in names

File: front/src/components/Schedule/index.js [hunk 20]
```diff
@@ -7,9 +7,14 @@ import { translate } from 'react-i18next';
 import { getLabById, getLecturesByLabId, toggleLabActivityState } from '../../services/labs';
 import { getFormattedDate } from '../../utils/datetime';
 import './Schedule.scss';
+import AdminFeedback from '../AdminFeedback';
+import MentorFeedback from '../MentorFeedback';
+
 
 class Schedule extends Component {
 
+  submenuContent = '';
```
Comment (AndreiShenets) [inline front/src/components/Schedule/index.js]: What is it for?

File: front/src/services/feedbacks.js [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+        'method': method,
+        'headers': this.headers
+      };
+    };
+
+    getUpsertDateOptions = (date, labId, id) => {
+      return {
+        'method': 'PUT',
+        'body': JSON.stringify({
+          'date': date,
```
Comment (AndreiShenets) [inline front/src/services/feedbacks.js]: quotes in names

File: front/src/services/fetchService/redirectAwareFetch.js [hunk 22]
```diff
@@ -7,23 +7,23 @@ function redirectAwareFetch(url, option, isCalledFromAuthentication) {
...
           return result;
         } else {
           window.location.pathname = '/authentication';
           break;
         }
 
-      case FetchResultTypeEnum.INTERNAL_SERVER_ERROR:
-        window.location.pathname = '/error';
+      case HttpStatus.INTERNAL_SERVER_ERROR:
+        // window.location.pathname = '/error';
```
Comment (AndreiShenets) [inline front/src/services/fetchService/redirectAwareFetch.js]: Why is this commented?

File: ITechArt.StudentsLab.BusinessLayer/Models/UpsertResponseModel.cs [hunk 23]
```diff
@@ -0,0 +1,18 @@
+﻿namespace ITechArt.StudentsLab.BusinessLayer.Models
+{
+    public class UpsertResponseModel
```
Comment (AndreiShenets) [inline ITechArt.StudentsLab.BusinessLayer/Models/UpsertResponseModel.cs]: Why do you mention response on BL layer?

File: ITechArt.StudentsLab.BusinessLayer/Models/UpsertResponseModel.cs [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public DalUpsertExecutionResult Result { get; }
```
Comment (AndreiShenets) [inline ITechArt.StudentsLab.BusinessLayer/Models/UpsertResponseModel.cs]: Why do you use DAL in BL model?
